File: src/platform/api/updater.ts
import { isUpdaterStatus } from '../ipcContracts.ts';
import type { EventRegistrar, InvokeFn, UpdaterStatus } from '../ipcContracts.ts';

export function createUpdaterApi(invoke: InvokeFn, onEvent: EventRegistrar) {
  return {
    status: async (): Promise<UpdaterStatus | null> => {
      const status = await invoke('updater_status');
      return isUpdaterStatus(status) ? status : null;
    },
    check: () => invoke('updater_check'),
    download: () => invoke('updater_download'),
    install: () => invoke('updater_install'),
    onStatus: onEvent('updater:status', isUpdaterStatus),
  };
}

File: src-tauri/src/commands/updater.rs
use crate::ipc::{self, CommandError, CommandResult, ErrorCode, OkResult};
use crate::runtime::updater::{self, UpdaterState, UpdaterStatus};
use tauri::{AppHandle, State};

fn outcome(result: anyhow::Result<()>) -> OkResult {
    match result {
        Ok(()) => ipc::ok(),
        Err(error) => OkResult::Err {
            ok: false,
            error: CommandError::from_anyhow(&error),
        },
    }
}

/// What the updater last reported. The startup check begins with the
/// process, so it has often said something before the UI subscribed.
#[tauri::command]
pub fn updater_status(state: State<'_, UpdaterState>) -> Option<UpdaterStatus> {
    state.status()
}

#[tauri::command]
pub async fn updater_check(app: AppHandle) -> CommandResult<OkResult> {
    if !updater::updates_enabled() {
        updater::report_unavailable(&app);
        return Ok(OkResult::Err {
            ok: false,
            error: CommandError::new(ErrorCode::Internal, "Updater unavailable in development"),
        });
    }
    Ok(outcome(updater::check(&app).await))
}

#[tauri::command]
pub async fn updater_download(app: AppHandle) -> CommandResult<OkResult> {
    Ok(outcome(updater::download(&app).await))
}

/// Shuts FTPeach down and hands over to the installer. It answers only when
/// there is no verified download to install.
#[tauri::command]
pub async fn updater_install(app: AppHandle) -> CommandResult<OkResult> {
    Ok(outcome(updater::install_now(&app).await))
}
